### PR TITLE
refactor(envoy): refactoring and optimising the components that build envoy config

### PR DESCRIPTION
**Pull request by driev**
**What this PR does / why we need it**:

This aims to simplify the envoy incremental code, by removing some iteration and increasing the "locality" of function calls. 

**Which issue(s) this PR fixes**:

Fixes # INFRA-1420

TODO: 

 - [x] merge https://github.com/SeldonIO/seldon-core/pull/6121 first

### CODE REVIEW

File: scheduler/pkg/envoy/xdscache/seldoncache.go [hunk 1]
```diff
@@ -352,27 +323,37 @@ func (xds *SeldonXDSCache) RemoveRoute(routeName string) error {
...
 
-	xds.Clusters[clusterName] = cluster
+	grpcCluster, ok := xds.Clusters[cluster.GrpcCluster]
+	if !ok {
+		return fmt.Errorf("can't find grpc cluster for route %s cluster %s route %+v", route.RouteName, cluster.GrpcCluster, route)
+	}
+	delete(grpcCluster.Routes, resources.RouteVersionKey{RouteName: route.RouteName, ModelName: cluster.ModelName, Version: cluster.ModelVersion})
+	if len(grpcCluster.Routes) == 0 {
+		delete(xds.Clusters, cluster.GrpcCluster)
+	}
```
Comment (sakoush) [inline scheduler/pkg/envoy/xdscache/seldoncache.go:357]: can this be refactor into a helper used by http and grpc?

**Review by sakoush — APPROVED** [review]
LGTM. some minor comments.

File: scheduler/pkg/envoy/xdscache/seldoncache.go [hunk 2]
```diff
@@ -36,7 +37,8 @@ const (
 )
 
 type SeldonXDSCache struct {
-	Listeners              map[string]resources.Listener
+	permanentListeners     []types.Resource
```
[the remark this answers is not in the document]
Comment (driev) [reply scheduler/pkg/envoy/xdscache/seldoncache.go:40]: So they don't need to be recreated on every update to envoy.

File: scheduler/pkg/envoy/xdscache/seldoncache.go [hunk 3]
```diff
@@ -54,18 +56,56 @@ type PipelineGatewayDetails struct {
 
 func NewSeldonXDSCache(logger logrus.FieldLogger, pipelineGatewayDetails *PipelineGatewayDetails) *SeldonXDSCache {
 	return &SeldonXDSCache{
-		Listeners:              make(map[string]resources.Listener),
+		permanentListeners:     make([]types.Resource, 2),
+		permanentClusters:      make([]types.Resource, 4),
```
[the remark this answers is not in the document]
Comment (driev) [reply scheduler/pkg/envoy/xdscache/seldoncache.go:60]: I'll move them to a constant.

File: scheduler/pkg/envoy/xdscache/seldoncache.go (hunk 1 above)
Comment (driev) [reply scheduler/pkg/envoy/xdscache/seldoncache.go:357]: Yes.

File: scheduler/pkg/envoy/resources/resource.go [hunk 4]
```diff
@@ -565,156 +684,46 @@ func makePipelineStickySessionRoute(r *PipelineRoute, clusterTraffic *PipelineTr
...
-	return len(r.Clusters) > 1
-}
-
-func calcNumberOfModelStickySessionsNeeded(modelRoutes []*Route) int {
+func countModelStickySessions(routes map[string]Route) int {
 	count := 0
-	for _, r := range modelRoutes {
-		if isModelExperiment(r) {
+	for _, r := range routes {
+		if len(r.Clusters) > 1 {
```
[the remark this answers is not in the document]
Comment (driev) [reply scheduler/pkg/envoy/resources/resource.go:690]: Added it back, for now.. I was going to make the model more explicit.

**Pull request merged**